services/acm/src/main/java/software/amazon/awssdk/services/acm/auth/scheme/internal/AcmAuthSchemeInterceptor.java

```diff
@@ -0,0 +1,132 @@
+
+    private static final Logger log = Logger.loggerFor(AcmAuthSchemeInterceptor.class);
+
+    @Override
+    public void beforeExecution(Context.BeforeExecution context, ExecutionAttributes executionAttributes) {
+        List<AuthSchemeOption> authOptions = resolveAuthOptions(executionAttributes);
+        SelectedAuthScheme<?> selectedAuthScheme = selectAuthScheme(authOptions, executionAttributes);
+        executionAttributes.putAttribute(SdkExecutionAttribute.SELECTED_AUTH_SCHEME, selectedAuthScheme);
+    }
+
+    /**
+     * Invoke the auth scheme resolver to determine which auth options we should consider for this request.
+     */
+    private List<AuthSchemeOption> resolveAuthOptions(ExecutionAttributes executionAttributes) {
+        // Prepare the inputs for the auth scheme resolver. We always include the
+        // operationName, and we include the region if the service is modeled with
+        // @sigv4.
+        String operation = executionAttributes.getAttribute(SdkExecutionAttribute.OPERATION_NAME);
+        // TODO: Do this only if sigv4
+        Region region = executionAttributes.getAttribute(AwsExecutionAttribute.AWS_REGION);
+
+        // Use the operation-level auth scheme resolver if the customer
+        // specified an override, otherwise fall back to the one on the client.
+        // TODO: The fallback should happen while setting in execution attributes
+        AcmAuthSchemeProvider authSchemeProvider =
+            (AcmAuthSchemeProvider) executionAttributes.getAttribute(SdkExecutionAttribute.AUTH_SCHEME_RESOLVER);
+
+        return authSchemeProvider.resolveAuthScheme(AcmAuthSchemeParams.builder()
+                                                                       .operation(operation)
+                                                                       .region(region)
+                                                                       .build());
+    }
+
+    /**
+     * From a list of possible auth options for this request, determine which auth scheme should be used.
+     */
+    private SelectedAuthScheme<?> selectAuthScheme(List<AuthSchemeOption> authOptions, ExecutionAttributes executionAttributes) {
+
+        // TODO: This should be "merged" earlier, with request preferred over client
+        Map<String, AuthScheme<?>> authSchemes =
+            executionAttributes.getAttribute(SdkExecutionAttribute.AUTH_SCHEMES);
+
+        IdentityProviderConfiguration identityResolvers =
+            executionAttributes.getAttribute(SdkExecutionAttribute.IDENTITY_PROVIDER_CONFIGURATION);
+
+        StringBuilder failureReasons = new StringBuilder();
+
+        // Check each option, in the order the auth scheme resolver proposed them.
+        for (AuthSchemeOption authOption : authOptions) {
+            // If we're using no-auth, don't consider which options are enabled.
+            if (authOption.schemeId().equals("smithy.auth#noAuth")) {
```

Comment:
Is this synthetic ID definitely approved?

Reply:
Not yet. I still have to work on getting that reviewed.